bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tools/watcher/EimJsonUiChangeHandler.java

```diff
@@ -32,13 +48,99 @@ public EimJsonUiChangeHandler(Preferences preferences)
 	@Override
 	public void onJsonFileChanged(Path file)
 	{
-		Display.getDefault().asyncExec(() -> {
-			MessageDialog.openWarning(Display.getDefault().getActiveShell(), Messages.EimJsonChangedMsgTitle,
-					Messages.EimJsonChangedMsgDetail);
+		int response = displayMessageToUser();
+		handleUserResponse(response);
+	}
+	
+	public int displayMessageToUser()
+	{
+		final int[] response = new int[] {-1};
+		Display display = Display.getDefault();
+		display.syncExec(() -> {
+			Shell shell = display.getActiveShell();
+			if (shell == null)
+			{
+				shell = new Shell(display);
+			}
+			MessageBox messageBox = new MessageBox(shell, SWT.ICON_WARNING | SWT.YES | SWT.NO);
+			messageBox.setText(Messages.EimJsonChangedMsgTitle);
+			messageBox.setMessage(Messages.EimJsonChangedMsgDetail);
+			response[0] = messageBox.open();
 		});
+		
+		return response[0];
+	}
 
+	public void handleUserResponse(int response)
+	{
+		if (response == SWT.YES)
+		{
+			try
+			{
+				loadEimJson();
+				if (eimJson.getIdfInstalled().size() == 1)
+				{
+					// only one entry in eimJson so we can simply refresh the IDE environment with that.
+					setupToolsInIde();
+				}
+				else 
+				{
+					// multiple entries in json so launch manager for user to handle this
+					launchEspIdfManager();
+				}
+			}
+			catch (IOException | PartInitException e)
+			{
+				Logger.log(e);
+			}
+		}
+		
 		EimJsonStateChecker checker = new EimJsonStateChecker(preferences);
 		checker.updateLastSeenTimestamp();
 	}
+	
+	private void loadEimJson() throws IOException
+	{
+		EimIdfConfiguratinParser eimIdfConfiguratinParser = new EimIdfConfiguratinParser();
+		eimJson = eimIdfConfiguratinParser.getEimJson(true);
+	}
+	
+	
+	
+	private void setupToolsInIde()
+	{
+		SetupToolsInIde setupToolsInIde = new SetupToolsInIde(eimJson.getIdfInstalled().get(0), eimJson, getConsoleStream(true), getConsoleStream(false));
+		setupToolsInIde.schedule();
+	}
+	
+	private void launchEspIdfManager() throws PartInitException
+	{
+		Display.getDefault().asyncExec(() -> {
+	        IWorkbenchWindow activeww = EclipseHandler.getActiveWorkbenchWindow();
+	        if (activeww != null && activeww.getActivePage() != null)
+	        {
+	            try
+	            {
+	                IDE.openEditor(activeww.getActivePage(), new EimEditorInput(eimJson),
+	                        ESPIDFManagerEditor.EDITOR_ID, true);
+	            }
+	            catch (PartInitException e)
+	            {
+	                Logger.log("Failed to open ESP-IDF Manager Editor.");
+	                Logger.log(e);
+	            }
+	        }
+	        else
+	        {
+	            Logger.log("Cannot open ESP-IDF Manager Editor. No active workbench window yet.");
+	        }
+	    });
```

Comment:
Avoid nested blocks
If certain criteria doesn't match, verify and return immedialy

```
if (activeww == null || activeww.getActivePage() == null) {
                Logger.log("Cannot open ESP-IDF Manager Editor. No active workbench window or page.");
                return;
            }
```

Reply:
Done